Finish Typescript Migration

herbertli opened the pull request:
Migrated rest of the files to typescript @yoav-zibin @apsdehal

herbertli committed:
Migrate withDragDropContext.js and Fix Merge in Board

- Added types for react-dnd-html5-backend
- Fixed Board.tsx

herbertli committed:
Migrate SpecViewer

- Added typings

herbertli committed:
Migrate ElementCreator and ElementCreatorContainer

- Add typings to ElementCreatorContainer.tsx
- Add typings to ElementCreator.tsx
- Change some props in Element.tsx to be optional
- Add null checks in Element.tsx

herbertli committed:
Migrate ImageSelector

- Added typings

herbertli committed:
Migrate ImageUploader

- Added typings to ImageUploader
- Made one of props in ImageSelector optional

herbertli committed:
Migrate ElementKindSelector

- Added typings

herbertli committed:
Migrate Login

- Added typings

herbertli committed:
Migrate GameSpecBuilder

- Added types
- Added props to Board.tsx

herbertli committed:
Migrate GameSpecBuilderContainer

- Added types
- Added props to SpecInfo
- Added props to SpecViewer

herbertli committed:
Migrate PlaySpecContainer

- Added types

herbertli committed:
Migrate SpecTestContainer

- Added types

herbertli committed:
Migrate registerServiceWorker

- Add types

herbertli committed:
Migrate RouteWhenAuthenticated

- Added types
- Got rid of extranneous props

herbertli committed:
Migrate App

- Added types
- Added/fixed props to NavigationMaster

apsdehal commented on src/components/ImageUploader.tsx:
```diff
@@ -13,40 +13,62 @@ import styles from '../styles';
+
+class ImageUploader extends React.Component<ImageUploaderProps, ImageUploaderState> {
+
+  initialState: any;
```
Can't initial state's type be set to ImageUploaderState you created above?

apsdehal commented on src/components/ImageUploader.tsx:
```diff
@@ -13,40 +13,62 @@ import styles from '../styles';
     return new Promise((resolve, reject) => {
-      window.URL = window.URL || window.webkitURL;
+
+      // window.URL = window.URL || window.webkitURL;
```
Any specific reason to uncomment this?

apsdehal commented on src/components/ImageUploader.tsx:
```diff
@@ -146,16 +177,16 @@ class ImageUploader extends React.Component {
+  }
 
-  handleImageUploaderChange = (element, e, newValue) => {
+  handleImageUploaderChange = (element: string, e: any, newValue: any) => {
```
`newValue` would be a string here.

apsdehal commented on src/components/SpecTestContainer.tsx:
```diff
@@ -17,54 +17,100 @@ import TextField from 'material-ui/TextField';
-
-  state = Object.assign({}, this.initialState, this.initialBoardState);
-  vars = Object.assign({}, this.initialVars);
+interface SpecTestContainerProps {
```
You can directly mention `{}` there, no need to create this.

apsdehal commented on src/components/SpecTestContainer.tsx:
```diff
@@ -17,54 +17,100 @@ import TextField from 'material-ui/TextField';
+
+class SpecTestContainer extends React.Component<SpecTestContainerProps, SpecTestContainerState> {
+
+  initialState: any;
```
Same as before. So many `any` use. Though I said you can use `any`, I think these one's should be type tested as these should not be ever set to wrong type of value, as it will create a lot of errors.

herbertli replied:
I've changed the type for other cases of Object.assign(), however, for this particular case the overall state is split up between initialState and initialBoardState. I could define two separate interfaces instead though.

apsdehal commented on src/components/SpecTestContainer.tsx:
```diff
@@ -375,29 +424,29 @@ class SpecTestContainer extends React.Component {
           <SpecInfo
-            handleIcon50CLick={this.handleGameIcon50.bind(this)}
-            handleIcon512Click={this.handleGameIcon512.bind(this)}
+            handleIcon50CLick={(key: any) => this.handleGameIcon50(key)}
```
`key` is number I think

apsdehal commented on src/components/gamespec/Board.tsx:
```diff
@@ -149,23 +151,12 @@ class Board extends React.Component<BoardProps, BoardState> {
     });
   }
 
-<<<<<<< HEAD:src/components/gamespec/Board.js
```
Oops, I forgot to save the file after resolving conflict. Thanks.

apsdehal commented on src/components/gamespec/Element.tsx:
```diff
@@ -57,6 +60,9 @@ class Element extends React.Component<ElementProps, ElementState> {
     let index = data.indexOf(key);
     data.splice(index, 1);
     this.setState({ selected_cards: data });
+    if (!this.props.setSelectedElements) {
```
Maybe check for both in the start itself. There is no point in reaching this step is `setSelectedElements` is undefined.

apsdehal commented on src/components/GameSpecBuilderContainer.tsx:
```diff
@@ -245,44 +289,47 @@ class GameSpecBuilderContainer extends React.Component {
   }
 
-  setWiki(e, newValue) {
+  setWiki(e: any, newValue: string) {
```
I think there should be standard type for `e`. Lookup if possible?

apsdehal reviewed (changes requested):
Some minor comments! Overall, a lot of work, good job!.

herbertli committed:
Added Types

- Added typings

herbertli committed:
Stricter Typing for ElementCreator, ElementCreatorContainer, Element

- Got rid of all the anys

herbertli committed:
Replaced Typings

- Added types for event handlers

herbertli committed:
Update dependencies

- updated yarn.lock

herbertli committed:
Fixes

ImageUploader
- Added types to event handlers
- Fixed webkit.url

Piece
- Added types

ElementTile/Board/BoardTile
- Delete some 'any's and replaced with proper types

Element
- Moved null check up

herbertli commented:
@apsdehal I think I've fixed most of the things, I also got rid of some of the 'any's and replaced them with the correct types

apsdehal commented:
This has landed into master. Thanks for your awesome work.
